log/sequencer.go

```diff
@@ -130,26 +130,30 @@ func NewSequencer(
 	}
 }
 
-// TODO: This currently doesn't use the batch api for fetching the required nodes. This
-// would be more efficient but requires refactoring.
 func (s Sequencer) buildMerkleTreeFromStorageAtRoot(ctx context.Context, root *types.LogRootV1, tx storage.TreeTX) (*compact.Tree, error) {
-	mt, err := compact.NewTreeWithState(s.hasher, int64(root.TreeSize), func(depth int, index int64) ([]byte, error) {
-		nodeID, err := storage.NewNodeIDForTreeCoords(int64(depth), index, maxTreeDepth)
-		if err != nil {
-			return nil, fmt.Errorf("%x: Failed to create nodeID: %v", s.signer.KeyHint, err)
+	mt, err := compact.NewTreeWithState(s.hasher, int64(root.TreeSize), func(ids []compact.NodeID) ([][]byte, error) {
+		storIDs := make([]storage.NodeID, len(ids))
+		for i, id := range ids {
+			nodeID, err := storage.NewNodeIDForTreeCoords(int64(id.Level), int64(id.Index), maxTreeDepth)
+			if err != nil {
+				return nil, fmt.Errorf("%x: failed to create nodeID: %v", s.signer.KeyHint, err)
+			}
+			storIDs[i] = nodeID
 		}
-		nodes, err := tx.GetMerkleNodes(ctx, int64(root.Revision), []storage.NodeID{nodeID})
 
+		nodes, err := tx.GetMerkleNodes(ctx, int64(root.Revision), storIDs)
 		if err != nil {
-			return nil, fmt.Errorf("%x: Failed to get Merkle nodes: %v", s.signer.KeyHint, err)
+			return nil, fmt.Errorf("%x: failed to get Merkle nodes: %v", s.signer.KeyHint, err)
 		}
-
-		// We expect to get exactly one node here
-		if nodes == nil || len(nodes) != 1 {
-			return nil, fmt.Errorf("%x: Did not retrieve one node while loading compact Merkle tree, got %#v for ID %v@%v", s.signer.KeyHint, nodes, nodeID.String(), root.Revision)
+		if got, want := len(nodes), len(storIDs); got != want {
+			return nil, fmt.Errorf("%x: failed to get %d nodes at rev %d, got %d", s.signer.KeyHint, want, root.Revision, got)
 		}
 
-		return nodes[0].Hash, nil
+		hashes := make([][]byte, len(nodes))
+		for i, node := range nodes {
+			hashes[i] = node.Hash
```

Comment:
Given the tightening of the `GetMerkleNodes` contract, I think it's probably worth asserting the assumption holds here - the result of it not holding would be catastrophic...

wdyt?

Reply:
Great idea. I think I will sleep better if I do this :)

Reply:
Done.